godot-core/src/obj/traits.rs

```diff
@@ -685,6 +685,40 @@ pub trait Singleton: GodotClass {
     fn singleton() -> Gd<Self>;
 }
 
+/// Trait for user-defined singleton classes in Godot.
+///
+/// There should be only one instance of each singleton class in the engine, accessible through [`singleton()`][Singleton::singleton].
+// For now exists mostly as a marker trait and a way to provide blanket implementation for `Singleton` trait.
+pub trait UserSingleton:
+    GodotClass + Bounds<Declarer = bounds::DeclUser, Memory = bounds::MemManual>
+{
+}
+
+impl<T> Singleton for T
+where
+    T: UserSingleton + Inherits<crate::classes::Object>,
+{
+    fn singleton() -> Gd<T> {
+        // Note: with safeguards enabled `singleton_unchecked` will panic if Singleton can't be retrieved.
+        #[cfg(not(safeguards_strict))]
+        {
+            // SAFETY: The caller must ensure that `class_name` corresponds to the actual class name of type `T`.
+            // This is always true for proc-macro derived user singletons.
+            unsafe {
+                crate::classes::singleton_unchecked(&<T as GodotClass>::class_id().to_string_name())
+            }
+        }
+        #[cfg(safeguards_strict)]
+        {
+            let class = <T as GodotClass>::class_id();
+            let singleton = crate::classes::Engine::singleton()
+                .get_singleton(&class.to_string_name())
+                .unwrap_or_else(|| panic!("Singleton {class} not found. User singleton must be registered under its class name."));
+            singleton.try_cast().unwrap_or_else(|obj| panic!("Downcasting of User Engine Singleton {class} failed. Expected: {class}, found: {}", obj.dynamic_class_string()))
```

Comment:
Formatting is not great, is it one of the cases where `rustfmt` gives up due to `panic!` being a macro?

Does it get better if you don't use an intermediate variable, but directly chain `try_cast()` after `unwrap_or_else()`?